x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/TransportGrantAction.java

```diff
@@ -73,12 +75,15 @@ protected void executeWithGrantAuthentication(GrantRequest grantRequest, ActionL
                     } else {
                         // Authentication can be run-as even when runAsUsername is null.
                         // This can happen when the authentication itself is a run-as client-credentials token.
-                        assert runAsUsername != null || "access_token".equals(grantRequest.getGrant().getType());
+                        assert runAsUsername != null || "access_token".equals(request.getGrant().getType());
                         authorizationService.authorize(
                             authentication,
                             AuthenticateAction.NAME,
                             new AuthenticateRequest(effectiveUsername),
-                            ActionListener.wrap(ignore2 -> listener.onResponse(authentication), listener::onFailure)
+                            ActionListener.wrap(
+                                ignore2 -> doExecuteOnGrantAuthenticationSuccess(task, request, authentication, listener),
```

Comment:
Nit: I slightly prefer `doExecuteWithGrantAuthentication` because it also does a bit authorization here so it is not just "AuthenticationSuccess".